### ISSUE TITLE
Support async code in dangerfile

### ISSUE THREAD

**Issue by kwonoj** [danger/danger-js#88]
How does async code works in dangerfile? I gave quick try like

```js
const fs = require("fs");
const os = require("os");
const summary = require("lcov-summary");

const file = fs.readFileSync("./coverage/lcov.info", "utf-8");

summary(file, (err, results) => {
  console.log('here');
  if (err) {
    warn("cannot retrieve code coverage");
  } else {
    console.log('there');
    const total = results
      .split(os.EOL)
      .filter((result) => result.match(/Total Coverage/))[0];
    message(total);
    warn(total);
    fail('fail');
  }
});
```

to write down simple message inside of callback. Callback's executed, but none of danger messages are written down. is this expected?

**Comment by orta** [comment]
Not expected, this should work - I've not touched any async code yet in my own dangerfiles. However, I would have thought that we would have got a notification of when the vm script finished, which would include all potential async callbacks wrapping up. 

Maybe that assumption is wrong

**Comment by kwonoj** [comment]
So, for this issue.. reason was quite simple, but I couldn't make a feasible solution to propose with current public interfaces.

What's happening inside of runner of danger imports dangerfile (https://github.com/danger/danger-js/blob/5ddce5937252bf7d915cb2cb40484e3ca7788444/source/runner/DangerfileRunner.ts#L73) and that's just it as whole execution and rely on all eval is done when module is being imported. It is naïvely working model, since when CJS module is being imported it'll be evaluated immediately (and statically), so all of synchronous danger execution will be evaluated as well. 

This becomes problematic with async code which starts evaluation but runner does not able to wait those async code to be completed. Runner may need to deliver some sort of way to get signal, and dangerfile need to explicitly signal once it's done.

**Comment by orta** [comment]
Ouch, that's not optimal!

I took a look around to see if I could find any existing implementations that handle this problem, and there are two:

* https://github.com/dallonf/async-eval
* and https://github.com/brianmcd/contextify (which comes with it's own "you do not need this dep" warning)

async-eval looks like it actually does what you're talking about by wrapping the eval'd code. 

Another potential place where we could find answers is inside [vm2](https://www.npmjs.com/package/vm2), which is a lib I considered instead of jest-runtime - their example uses an async script

_sidenote:_ @cpojer - is this something you've had to handle in Jest?

**Comment by cpojer** [comment]
the call returns as soon as the script is done with the execution of whatever you run in the top level (blocking). You'll need to do something like, require a module in the context that provides "danger" and then do something like `danger.schedule(asyncWork)`. Then you can do `danger.wait(…).then(…)` or something in your code. Does that make sense?

**Comment by orta** [comment]
That's the concept you use returning a promise in an `it` right?

**Comment by cpojer** [comment]
Yeah, pretty much, except that's all encapsulated somewhere in the shitshow that is jasmine.

**Comment by DaleJefferson** [comment]
How about an API similar to Jasmine / Jest.

```
// Done callback
test('coverage', (done) => {
    const file = fs.readFileSync("./coverage/lcov.info", "utf-8");
    
    summary(file, (err, results) => {
     ...
     warn()
     done() 
     });
})

// Promise
test('coverage', (done) => {
    const file = fs.readFileSync("./coverage/lcov.info", "utf-8");

    return summary(file).then((results) => {
        ....
        warn()
    })
})

// Async / Await
test('coverage', (done) => {
    const file = fs.readFileSync("./coverage/lcov.info", "utf-8");

    const results = await summary(file)
    ...
    warn()
})
```

This in theory would also allow messages to be generated in parallel.

**Comment by orta** [comment]
Yeah, I think this pattern is probably the right appraoch 👍

**Comment by jsdario** [comment]
Hi, I am new to this project, is there anyway we could help or this feature is not friendly for beginners?

**Comment by kwonoj** [comment]
if you have working ideas I'd welcome it. I've been thinking this while but didn't have great way to achieve to support this as well as not breaking existing interfaces of dangerfile as well.

**Comment by orta** [comment]
I'm OK with breaking the interfaces, I also wonder if there's something we can do on our side, it seems so strange to not be able to wait till all related ticks from our own process are finished

**Comment by orta** [comment]
Danger is still pre-1.0, and I probably need to do some foundation shaking for Peril anyway

**Comment by deecewan** [comment]
if we implement it in the way described in https://github.com/danger/danger-js/issues/88#issuecomment-275900823, there should be no break to the interface for existing dangerfiles.  Just, new dangerfiles will be able to use `schedule` (which I'm a fan of).

Of course, the terminology could change. If we go with `danger.schedule`, an example workflow could look like (and, I haven't looked at the result of the code parsing):

```js
// inside dangerfile.js
import { schedule, warn } from 'danger';

schedule(async () => {
  const res = await doSomethingAsync();
  if (res) {
    warn(res);
  }
});
```

```js
const asyncs = [];
const result = parseDangerfile();
// asyncs now contains [fn1, fn2, fn3];
Promise.all(asyncs);
```

I'd be keen to have a crack at this over the next couple of days.

**Comment by deecewan** [comment]
hmmm.  I've been playing with this for a while and have some questions for whoever:

- why are we using the `jest-runtime`?  What's stopping us from just requiring the code?
  - is this for transpilation?
- if we export a singleton, we don't need to modify the Dangerfile, right?  I think this is ideal, because if an error occurs during the running the file stays in the broken state.

I realise that this would probably be a pretty massive change, and I'm probably missing some subtleties.

**Comment by orta** [comment]
Jest runtime is there for transpilation, but also so that we can support evaling un-trusted code ( by using it's module mocking systems ) for things like hosted Danger ([Peril](https://github.com/danger/peril))

And yep, schedule would indeed work! :D

**Comment by deecewan** [comment]
i was having trouble last night getting access to the context to resolve the scheduled code.

i'll push up the code I wrote and link it here so you can see what I've done/find out why it doesn't work.

**Comment by deecewan** [comment]
This issue is mostly resolved by #144.  It does not support observables, but supports callbacks, async/await and promises.  Feel free to open a new issue if you'd like other kinds of async supported.
